File: services/flight-search/src/providers/skyscanner.rs
//! Skyscanner live-search adapter.
//!
//! Fetching (create + poll session flow) is kept separate from the pure
//! normalisation functions so tests can cover the mapping with fixture JSON.

use std::collections::BTreeMap;
use std::time::Duration;

use serde::Deserialize;
use serde_json::{json, Value};

use crate::aggregator::MAX_OFFERS;
use crate::config::SkyscannerConfig;
use crate::error::ProviderError;
use crate::models::{FlightOffer, FlightSearch};
use crate::providers::send_with_retry;

const MAX_POLL_ATTEMPTS: usize = 3;
const POLL_DELAY: Duration = Duration::from_secs(1);

/// Search Skyscanner live prices for a validated request.
///
/// `live/search/create` is session-based: the first response is usually
/// `RESULT_STATUS_INCOMPLETE` with few or no itineraries, so the session is
/// polled until it completes (or we run out of attempts).
pub async fn search(
    client: &reqwest::Client,
    config: &SkyscannerConfig,
    search: &FlightSearch,
) -> Result<Vec<FlightOffer>, ProviderError> {
    let body = build_create_body(search);
    let create_url = format!("{}/v3/flights/live/search/create", config.base_url);

    let response = send_with_retry(|| {
        client
            .post(&create_url)
            .header("x-api-key", &config.api_key)
            .json(&body)
    })
    .await?;
    if !response.status().is_success() {
        tracing::error!("skyscanner search failed: {}", response.status());
        return Ok(Vec::new());
    }

    let mut data: Value = response.json().await?;
    let session_token = data
        .get("sessionToken")
        .and_then(Value::as_str)
        .map(str::to_owned);

    if let Some(session_token) = session_token {
        let poll_url = format!(
            "{}/v3/flights/live/search/poll/{session_token}",
            config.base_url
        );
        for _ in 0..MAX_POLL_ATTEMPTS {
            if is_complete(&data) {
                break;
            }
            tokio::time::sleep(POLL_DELAY).await;
            let poll_response =
                send_with_retry(|| client.post(&poll_url).header("x-api-key", &config.api_key))
                    .await?;
            if !poll_response.status().is_success() {
                tracing::warn!("skyscanner poll failed: {}", poll_response.status());
                break;
            }
            data = poll_response.json().await?;
        }
    }

    Ok(normalize_response(
        &data,
        search,
        config.affiliate_id.as_deref(),
    ))
}

/// True when the session response reports `RESULT_STATUS_COMPLETE`.
pub fn is_complete(raw: &Value) -> bool {
    raw.pointer("/content/status").and_then(Value::as_str) == Some("RESULT_STATUS_COMPLETE")
}

/// Build the `live/search/create` request body.
pub fn build_create_body(search: &FlightSearch) -> Value {
    let mut query_legs = vec![json!({
        "originPlaceId": { "iata": search.origin },
        "destinationPlaceId": { "iata": search.destination },
        "date": date_parts(&search.departure_date),
    })];
    if let Some(return_date) = &search.return_date {
        query_legs.push(json!({
            "originPlaceId": { "iata": search.destination },
            "destinationPlaceId": { "iata": search.origin },
            "date": date_parts(return_date),
        }));
    }
    json!({
        "query": {
            "market": "UK",
            "locale": "en-GB",
            "currency": "GBP",
            "queryLegs": query_legs,
            "cabinClass": search.cabin_class.skyscanner_cabin_class(),
            "adults": search.passengers,
        }
    })
}

fn date_parts(date: &str) -> Value {
    let mut parts = date
        .splitn(3, '-')
        .map(|part| part.parse::<u32>().unwrap_or(0));
    json!({
        "year": parts.next().unwrap_or(0),
        "month": parts.next().unwrap_or(0),
        "day": parts.next().unwrap_or(0),
    })
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SkyscannerResponse {
    content: Option<SkyscannerContent>,
}

#[derive(Debug, Deserialize)]
struct SkyscannerContent {
    results: Option<SkyscannerResults>,
}

#[derive(Debug, Deserialize)]
struct SkyscannerResults {
    #[serde(default)]
    itineraries: BTreeMap<String, SkyscannerItinerary>,
    #[serde(default)]
    legs: BTreeMap<String, SkyscannerLeg>,
    #[serde(default)]
    carriers: BTreeMap<String, SkyscannerCarrier>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SkyscannerItinerary {
    #[serde(default)]
    pricing_options: Vec<PricingOption>,
    #[serde(default)]
    leg_ids: Vec<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PricingOption {
    #[serde(default)]
    price: PriceAmount,
    #[serde(default)]
    items: Vec<PricingItem>,
}

#[derive(Debug, Default, Deserialize)]
struct PriceAmount {
    amount: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct PricingItem {
    deep_link: Option<String>,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct SkyscannerLeg {
    departure_date_time: SkyscannerDateTime,
    arrival_date_time: SkyscannerDateTime,
    duration_in_minutes: u32,
    stop_count: u32,
    #[serde(default)]
    operating_carrier_ids: Vec<String>,
    #[serde(default)]
    marketing_carrier_ids: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct SkyscannerDateTime {
    year: i64,
    month: u32,
    day: u32,
    hour: u32,
    minute: u32,
}

#[derive(Debug, Deserialize)]
struct SkyscannerCarrier {
    name: Option<String>,
}

fn to_iso_date_time(dt: &SkyscannerDateTime) -> String {
    format!(
        "{:04}-{:02}-{:02}T{:02}:{:02}:00",
        dt.year, dt.month, dt.day, dt.hour, dt.minute
    )
}

fn airline_name(
    leg: &SkyscannerLeg,
    carriers: &BTreeMap<String, SkyscannerCarrier>,
) -> Option<String> {
    let carrier_id = leg
        .operating_carrier_ids
        .first()
        .or_else(|| leg.marketing_carrier_ids.first())?;
    carriers.get(carrier_id).and_then(|c| c.name.clone())
}

/// Append `associateid=<affiliate>` to a deep link when an affiliate id is
/// configured. Empty URLs pass through untouched.
pub fn append_affiliate_id(url: &str, affiliate_id: Option<&str>) -> String {
    let Some(affiliate_id) = affiliate_id.filter(|id| !id.is_empty()) else {
        return url.to_string();
    };
    if url.is_empty() {
        return url.to_string();
    }
    let separator = if url.contains('?') { '&' } else { '?' };
    format!(
        "{url}{separator}associateid={}",
        percent_encode(affiliate_id)
    )
}

fn percent_encode(s: &str) -> String {
    use std::fmt::Write as _;
    let mut out = String::with_capacity(s.len());
    for byte in s.bytes() {
        match byte {
            b'A'..=b'Z' | b'a'..=b'z' | b'0'..=b'9' | b'-' | b'_' | b'.' | b'~' => {
                out.push(byte as char);
            }
            _ => {
                let _ = write!(out, "%{byte:02X}");
            }
        }
    }
    out
}

/// Pure normalisation: a raw Skyscanner session response in, normalised offers
/// out. Itineraries whose legs or carriers cannot be resolved are skipped
/// rather than emitted with placeholder values. Capped at `MAX_OFFERS`.
pub fn normalize_response(
    raw: &Value,
    search: &FlightSearch,
    affiliate_id: Option<&str>,
) -> Vec<FlightOffer> {
    let parsed = match SkyscannerResponse::deserialize(raw) {
        Ok(parsed) => parsed,
        Err(err) => {
            tracing::warn!("skyscanner: unparseable response: {err}");
            return Vec::new();
        }
    };
    let Some(results) = parsed.content.and_then(|content| content.results) else {
        return Vec::new();
    };

    let mut offers = Vec::new();
    for (id, itinerary) in &results.itineraries {
        if offers.len() >= MAX_OFFERS {
            break;
        }

        let Some(outbound_leg) = itinerary
            .leg_ids
            .first()
            .and_then(|leg_id| results.legs.get(leg_id))
        else {
            continue;
        };
        let Some(airline) = airline_name(outbound_leg, &results.carriers) else {
            continue;
        };

        let pricing = itinerary.pricing_options.first();
        let price_gbp = pricing
            .and_then(|p| p.price.amount.as_deref())
            .and_then(|amount| amount.parse::<f64>().ok())
            .map(|amount| (amount * 100.0).round() as i64)
            .unwrap_or(0);
        let deep_link = pricing
            .and_then(|p| p.items.first())
            .and_then(|item| item.deep_link.clone())
            .unwrap_or_default();

        let mut offer = FlightOffer {
            id: format!("skyscanner-{id}"),
            provider: "skyscanner".to_string(),
            provider_ref: id.clone(),
            airline,
            airline_logo: None,
            flight_number: String::new(),
            origin: search.origin.clone(),
            destination: search.destination.clone(),
            departure_at: to_iso_date_time(&outbound_leg.departure_date_time),
            arrival_at: to_iso_date_time(&outbound_leg.arrival_date_time),
            duration_minutes: outbound_leg.duration_in_minutes,
            stops: outbound_leg.stop_count,
            stop_cities: Vec::new(),
            return_departure_at: None,
            return_arrival_at: None,
            return_duration_minutes: None,
            return_stops: None,
            price_gbp,
            currency: "GBP".to_string(),
            cabin_class: search.cabin_class.as_str().to_string(),
            deep_link: append_affiliate_id(&deep_link, affiliate_id),
            baggage_included: false,
        };

        if let Some(return_leg_id) = itinerary.leg_ids.get(1) {
            let Some(return_leg) = results.legs.get(return_leg_id) else {
                continue;
            };
            offer.return_departure_at = Some(to_iso_date_time(&return_leg.departure_date_time));
            offer.return_arrival_at = Some(to_iso_date_time(&return_leg.arrival_date_time));
            offer.return_duration_minutes = Some(return_leg.duration_in_minutes);
            offer.return_stops = Some(return_leg.stop_count);
        }

        offers.push(offer);
    }
    offers
}

File: apps/web/src/lib/flights/skyscanner.ts
import { logger } from "@/lib/logger";
import type { FlightOffer, FlightSearch } from "./types";

const FETCH_TIMEOUT_MS = 8_000;
const MAX_POLL_ATTEMPTS = 3;
const POLL_DELAY_MS = 1_000;
const MAX_OFFERS = 20;

/** Fetch with a timeout, retrying once on network error or a 5xx response. */
async function fetchWithRetry(url: string, init: RequestInit = {}): Promise<Response> {
  const attempt = () => fetch(url, { ...init, signal: AbortSignal.timeout(FETCH_TIMEOUT_MS) });
  try {
    const response = await attempt();
    return response.status >= 500 ? await attempt() : response;
  } catch {
    return attempt();
  }
}

function delay(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

export async function searchSkyscannerFlights(search: FlightSearch): Promise<FlightOffer[]> {
  const baseUrl =
    process.env.SKYSCANNER_BASE_URL ?? "https://partners.api.skyscanner.net/apiservices";
  const apiKey = process.env.SKYSCANNER_API_KEY;

  if (!apiKey) {
    logger.warn("Skyscanner API key not configured");
    return [];
  }

  const body = {
    query: {
      market: "UK",
      locale: "en-GB",
      currency: "GBP",
      queryLegs: [
        {
          originPlaceId: { iata: search.origin },
          destinationPlaceId: { iata: search.destination },
          date: parseDateParts(search.departureDate),
        },
        ...(search.returnDate
          ? [
              {
                originPlaceId: { iata: search.destination },
                destinationPlaceId: { iata: search.origin },
                date: parseDateParts(search.returnDate),
              },
            ]
          : []),
      ],
      cabinClass: mapCabinClass(search.cabinClass),
      adults: search.passengers,
    },
  };

  const createResponse = await fetchWithRetry(`${baseUrl}/v3/flights/live/search/create`, {
    method: "POST",
    headers: {
      "x-api-key": apiKey,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(body),
  });

  if (!createResponse.ok) {
    logger.error(`Skyscanner search failed: ${createResponse.status}`);
    return [];
  }

  let data = (await createResponse.json()) as SkyscannerResponse;
  const sessionToken = data.sessionToken;

  // live/search/create is session-based: the first response is usually
  // RESULT_STATUS_INCOMPLETE with few or no itineraries, so poll the session
  // until it completes (or we run out of attempts).
  for (
    let attempt = 0;
    attempt < MAX_POLL_ATTEMPTS &&
    sessionToken &&
    data.content?.status !== "RESULT_STATUS_COMPLETE";
    attempt++
  ) {
    await delay(POLL_DELAY_MS);
    const pollResponse = await fetchWithRetry(
      `${baseUrl}/v3/flights/live/search/poll/${sessionToken}`,
      {
        method: "POST",
        headers: { "x-api-key": apiKey },
      },
    );
    if (!pollResponse.ok) {
      logger.warn(`Skyscanner poll failed: ${pollResponse.status}`);
      break;
    }
    data = (await pollResponse.json()) as SkyscannerResponse;
  }

  return normaliseSkyscannerResults(data, search);
}

function appendAffiliateId(url: string): string {
  const affiliateId = process.env.SKYSCANNER_AFFILIATE_ID;
  if (!affiliateId || !url) return url;
  const separator = url.includes("?") ? "&" : "?";
  return `${url}${separator}associateid=${encodeURIComponent(affiliateId)}`;
}

function parseDateParts(dateStr: string): { year: number; month: number; day: number } {
  const [year, month, day] = dateStr.split("-").map(Number);
  return { year, month, day };
}

function mapCabinClass(cabin: string): string {
  const map: Record<string, string> = {
    economy: "CABIN_CLASS_ECONOMY",
    premium_economy: "CABIN_CLASS_PREMIUM_ECONOMY",
    business: "CABIN_CLASS_BUSINESS",
    first: "CABIN_CLASS_FIRST",
  };
  return map[cabin] ?? "CABIN_CLASS_ECONOMY";
}

type SkyscannerDateTime = {
  year: number;
  month: number;
  day: number;
  hour: number;
  minute: number;
};

type SkyscannerLeg = {
  departureDateTime: SkyscannerDateTime;
  arrivalDateTime: SkyscannerDateTime;
  durationInMinutes: number;
  stopCount: number;
  operatingCarrierIds?: string[];
  marketingCarrierIds?: string[];
};

type SkyscannerCarrier = {
  name?: string;
};

type SkyscannerItinerary = {
  pricingOptions: Array<{ price: { amount?: string }; items: Array<{ deepLink?: string }> }>;
  legIds: string[];
};

type SkyscannerContent = {
  status?: string;
  results?: {
    itineraries?: Record<string, SkyscannerItinerary>;
    legs?: Record<string, SkyscannerLeg>;
    carriers?: Record<string, SkyscannerCarrier>;
  };
};

type SkyscannerResponse = {
  sessionToken?: string;
  content?: SkyscannerContent;
};

function toIsoDateTime(dt: SkyscannerDateTime): string {
  const pad = (n: number) => String(n).padStart(2, "0");
  return `${dt.year}-${pad(dt.month)}-${pad(dt.day)}T${pad(dt.hour)}:${pad(dt.minute)}:00`;
}

function airlineName(
  leg: SkyscannerLeg,
  carriers: Record<string, SkyscannerCarrier>,
): string | null {
  const carrierId = leg.operatingCarrierIds?.[0] ?? leg.marketingCarrierIds?.[0];
  if (!carrierId) return null;
  return carriers[carrierId]?.name ?? null;
}

function normaliseSkyscannerResults(data: SkyscannerResponse, search: FlightSearch): FlightOffer[] {
  const results = data.content?.results;
  const itineraries = results?.itineraries ?? {};
  const legs = results?.legs ?? {};
  const carriers = results?.carriers ?? {};

  const offers: FlightOffer[] = [];
  for (const [id, itinerary] of Object.entries(itineraries)) {
    if (offers.length >= MAX_OFFERS) break;

    // Skip itineraries whose legs/carriers cannot be resolved rather than
    // emitting "Various"/zero placeholders.
    const outboundLeg = legs[itinerary.legIds[0] ?? ""];
    if (!outboundLeg) continue;
    const airline = airlineName(outboundLeg, carriers);
    if (!airline) continue;

    const pricing = itinerary.pricingOptions[0];
    const offer: FlightOffer = {
      id: `skyscanner-${id}`,
      provider: "skyscanner",
      providerRef: id,
      airline,
      flightNumber: "",
      origin: search.origin,
      destination: search.destination,
      departureAt: toIsoDateTime(outboundLeg.departureDateTime),
      arrivalAt: toIsoDateTime(outboundLeg.arrivalDateTime),
      durationMinutes: outboundLeg.durationInMinutes,
      stops: outboundLeg.stopCount,
      stopCities: [],
      priceGBP: Math.round(Number(pricing?.price.amount ?? 0) * 100),
      currency: "GBP",
      cabinClass: search.cabinClass,
      deepLink: appendAffiliateId(pricing?.items[0]?.deepLink ?? ""),
      baggageIncluded: false,
    };

    const returnLegId = itinerary.legIds[1];
    if (returnLegId) {
      const returnLeg = legs[returnLegId];
      if (!returnLeg) continue;
      offer.returnDepartureAt = toIsoDateTime(returnLeg.departureDateTime);
      offer.returnArrivalAt = toIsoDateTime(returnLeg.arrivalDateTime);
      offer.returnDurationMinutes = returnLeg.durationInMinutes;
      offer.returnStops = returnLeg.stopCount;
    }

    offers.push(offer);
  }

  return offers;
}
